### PR TITLE
Add caching producer factory

### LINKED ISSUE #30: Allow producer factory to cache producers
The current DefaultProducerFactory does not cache the producers that it hands out. An improvement would be to reasonably cache the producers.

- Choice of cache?
  - caffeine
  - simple map ([example](https://github.com/spring-cloud/spring-cloud-stream/blob/6be9c461a541a6683b21fd42500c1cbed4b08b3f/spring-cloud-stream/src/main/java/org/springframework/cloud/stream/function/StreamBridge.java#L109))

### PR DESCRIPTION
**Pull request by onobc**
Fixes #30

### CODE REVIEW

File: spring-pulsar-boot-autoconfigure/src/main/java/org/springframework/pulsar/autoconfigure/PulsarAnnotationDrivenConfiguration.java [hunk 1]
```diff
@@ -70,7 +70,7 @@ public PulsarAnnotationDrivenConfiguration(PulsarProperties pulsarProperties) {
 	@Configuration(proxyBeanMethods = false)
 	@EnablePulsar
 	@ConditionalOnMissingBean(name = PulsarListenerBeanNames.PULSAR_LISTENER_ANNOTATION_PROCESSOR_BEAN_NAME)
-	static class EnableKafkaConfiguration {
+	static class EnablePulsarConfiguration {
```
Comment (onobc) [inline spring-pulsar-boot-autoconfigure/src/main/java/org/springframework/pulsar/autoconfigure/PulsarAnnotationDrivenConfiguration.java:73]: **[UNRELATED]** cleanup

File: spring-pulsar-boot-autoconfigure/src/main/java/org/springframework/pulsar/autoconfigure/PulsarProperties.java [hunk 2]
```diff
@@ -535,6 +542,43 @@ public Map<String, Object> buildProperties() {
 
 			return properties;
 		}
+
+	}
+
+	public static class Cache {
+
+		/** Time period to expire unused entries in the cache. */
+		private Duration expireAfterAccess = Duration.ofMinutes(1);
```
Comment (onobc) [inline spring-pulsar-boot-autoconfigure/src/main/java/org/springframework/pulsar/autoconfigure/PulsarProperties.java:551]: I left out any validation (`@NonNegative` etc...) as Caffeine is the final arbiter and does a very solid job of validating the inputs.

File: spring-pulsar-boot-autoconfigure/src/test/java/org/springframework/pulsar/autoconfigure/PulsarAutoConfigurationTests.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.autoconfigure;
+
+import static org.assertj.core.api.Assertions.assertThat;
```
Comment (onobc) [inline spring-pulsar-boot-autoconfigure/src/test/java/org/springframework/pulsar/autoconfigure/PulsarAutoConfigurationTests.java:19]: Ok so this is WEIRD.... the checkstyle has the following:
```xml
<module name="ImportOrder">
  <property name="groups" value="java,/^javax?\./,org,org.springframework,*"/>
  <property name="ordered" value="true"/>
  <property name="separated" value="true"/>
  <property name="option" value="top"/>
  <property name="sortStaticImportsAlphabetically" value="true"/>
</module>
```

It says `sortStaticImportsAlphabetically` but that is not what it does... it requires them to be at the very top before any other imports. It took me a good 30mins fighting this. Funny enough, IDEA using Spring Boot build conventions attempts to group the statics at the bottom and that is where they seem to make the most sense to me. Anyways... if you run into this @sobychacko please remember this strangeness and save 30mins of your life **NOT** batting the `ImportOrder` checkstyle failure.

File: spring-pulsar-boot-autoconfigure/src/test/java/org/springframework/pulsar/autoconfigure/PulsarAutoConfigurationTests.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+import org.springframework.pulsar.config.PulsarListenerEndpointRegistry;
+import org.springframework.pulsar.core.CachingPulsarProducerFactory;
+import org.springframework.pulsar.core.DefaultPulsarProducerFactory;
+import org.springframework.pulsar.core.PulsarConsumerFactory;
+import org.springframework.pulsar.core.PulsarProducerFactory;
+import org.springframework.pulsar.core.PulsarTemplate;
+import org.springframework.pulsar.listener.DefaultPulsarMessageListenerContainer;
+
+/**
+ * Autoconfiguration tests for {@link PulsarAutoConfiguration}.
```
Comment (onobc) [inline spring-pulsar-boot-autoconfigure/src/test/java/org/springframework/pulsar/autoconfigure/PulsarAutoConfigurationTests.java:47]: This covers the producer factory auto-config as well as every other possible aspect of auto-config (at the time of writing).

File: spring-pulsar/src/main/java/org/springframework/pulsar/config/PulsarListenerBeanNames.java [hunk 5]
```diff
@@ -28,12 +28,12 @@ public abstract class PulsarListenerBeanNames {
 	 * The bean name of the internally managed Pulsar listener annotation processor.
 	 */
 	public static final String PULSAR_LISTENER_ANNOTATION_PROCESSOR_BEAN_NAME =
-			"org.springframework.pulsar.config.internalKafkaListenerAnnotationProcessor";
+			"org.springframework.pulsar.config.internalPulsarListenerAnnotationProcessor";
```
Comment (onobc) [inline spring-pulsar/src/main/java/org/springframework/pulsar/config/PulsarListenerBeanNames.java:31]: **[UNRELATED]** cleanup

File: spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java [hunk 6]
```diff
@@ -76,28 +74,12 @@ public CompletableFuture<MessageId> sendAsync(String topic, T message, MessageRo
 						this.logger.error(ex, () -> String.format("Failed to send msg to '%s' topic", topicName));
 						// TODO fail metrics
 					}
-					closeProducerAsync(producer);
+					ProducerUtils.closeProducerAsync(producer, this.logger);
```
Comment (onobc) [inline spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java:77]: I am going to follow on w/ another update to PulsarTemplateTests if need be to ensure that close is being called now that this has been added

File: src/checkstyle/checkstyle.xml [hunk 7]
```diff
@@ -148,13 +148,6 @@
 					  value="Line has leading space characters; indentation should be performed with tabs only."/>
 			<property name="ignoreComments" value="true"/>
 		</module>
-		<module name="RegexpSinglelineJava">
-			<property name="maximum" value="0"/>
-			<property name="format" value="org\.mockito\..*Mockito\.(when|doThrow|doAnswer)"/>
-			<property name="message"
-					  value="Please use BDDMockito instead of Mockito.(when|doThrow|doAnswer)."/>
```
Comment (onobc) [inline src/checkstyle/checkstyle.xml:155]: If you don't mind I would like to remove the forced use of `BDDMockito`. This is something in most Spring projects that I do not understand the need for. Mockito (vanilla) has always been enough for me to do what I need and then some. Thoughts?

Comment (sobychacko) [reply src/checkstyle/checkstyle.xml:155]: Ya, I agree with that assessment. I just followed what other projects are doing. But Mockito is sufficient enough for our testing.

**Review by sobychacko — APPROVED** [review]

**Comment by sobychacko** [comment]
Thanks for this excellent PR @onobc. Significant improvements for `PulsarTemplate`. Merged upstream.

**Pull request merged**